spacy/language.py

```diff
@@ -1615,6 +1615,11 @@ def from_config(
                 or lang_cls is not cls
             ):
                 raise ValueError(Errors.E943.format(value=type(lang_cls)))
+        # Warn about require_gpu if a jupyter notebook + cupy is detected
+        if is_in_jupyter():
+            from thinc.backends.cupy_ops import CupyOps
+            if CupyOps.xp is not None:
+                warnings.warn(Warnings.W111)
```

Comment:
I don't really love the idea of just displaying this warning always, no matter what the user does. We should try to only warn if the user gets it wrong.

How about this:

* We set a global variable reflecting that the GPU has been enabled in some context. This global isn't going to work with threads or contexts, it's not the "source of truth". But it does tell us that something was done.
* When we load, we check whether the following is true: a) we're in a Jupyter notebook, b) The "use gpu" global is set, c) The "use gpu" context var isn't set. If all of those are true we should know we have a problem and can warn, right?

Reply:
Should we move all this logic into thinc, or do you just want to handle common spacy misusage here?

Reply:
We should move a lot of the logic into Thinc, but I don't see how to detect the different cells between `require_gpu` and `spacy.load` only in Thinc.

Reply:
See: https://github.com/explosion/thinc/pull/469

Reply:
No, I don't think we can detect what you've written, just that the basic thread stuff doesn't match the basic contextvars stuff. Obviously there are environments where this check wouldn't work, but it seems okay for the `is_in_jupyter` + `cupy` context?